File: 1391-check-if-there-is-a-valid-path-in-a-grid/1391-check-if-there-is-a-valid-path-in-a-grid.cpp
class Solution {
public:
    bool hasValidPath(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();

        vector<vector<int>> connects = {
            {}, {0,1}, {2,3}, {0,3}, {1,3}, {0,2}, {1,2},
        };

        vector<int> dr = {0,0,-1,1};
        vector<int> dc = {-1,1,0,0};
        vector<int> opp = {1,0,3,2};

        auto hasDir = [&](int type, int dir) {
            for (int d : connects[type]) if (d == dir) return true;
            return false;
        };

        vector<vector<bool>> visited(m, vector<bool>(n, false));
        queue<pair<int,int>> q;
        q.push({0, 0});
        visited[0][0] = true;
        
        while (!q.empty()) {
            auto [r, c] = q.front(); q.pop();
            if (r == m-1 && c == n-1) return true;
            
            for (int d = 0; d < 4; d++) {
                int nr = r + dr[d], nc = c + dc[d];
                if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
                if (visited[nr][nc]) continue;
                if (hasDir(grid[r][c], d) && hasDir(grid[nr][nc], opp[d])) {
                    visited[nr][nc] = true;
                    q.push({nr, nc});
                }
            }
        }
        return false;
    }
};

File: 1391-check-if-there-is-a-valid-path-in-a-grid/1391-check-if-there-is-a-valid-path-in-a-grid.py
class Solution(object):
    def hasValidPath(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: bool
        """
        m, n = len(grid), len(grid[0])
        connects = {
            1: {0, 1},
            2: {2, 3},
            3: {0, 3},
            4: {1, 3},
            5: {0, 2},
            6: {1, 2},
        }
        dr = [0, 0, -1, 1]
        dc = [-1, 1, 0, 0]
        opp = [1, 0, 3, 2]
        
        visited = [[False]*n for _ in range(m)]
        queue = [(0, 0)]
        visited[0][0] = True
        
        while queue:
            r, c = queue.pop(0)
            if r == m-1 and c == n-1:
                return True
            for d in range(4):
                nr, nc = r + dr[d], c + dc[d]
                if 0 <= nr < m and 0 <= nc < n and not visited[nr][nc]:
                    if d in connects[grid[r][c]] and opp[d] in connects[grid[nr][nc]]:
                        visited[nr][nc] = True
                        queue.append((nr, nc))
        return False